### PR TITLE
ProfileData API

### PR DESCRIPTION
**Pull request by Jongy**
## Description
This PR introduces a new API between the runtime profilers and gprofiler core: `ProfileData`.

APIs are not finalized yet.:
* [x] I don't like the current mess in `merge.py`
* [x] If appids / metadata are disabled, then we need to prevent gprofiler from trying to retrieve them (not just from adding them; with this PR, we still collect them despite not adding)

The idea is to use a more rich output format for profilers - not just PID -> stackcollapse, but PID -> object with numerous fields, one of them being stackcollapse.

Currently, the 2 additional fields are the appid and app metadata. Logically, it makes sense that the runtime profilers retrieve them, and not the generic area in gprofiler - because these are per-runtime anyway. Code is structured more "sensibly" when runtime profilers return these values.

Also, it serves a technical purpose: if profilers collect & attach the appid & app metadata directly to the profile data, we can be sure that:
* App id & metadata are always collected consistently - if a certain app has "appid: X", then all stacks of it will have that appid. Previously it wasn't the case, because we could profile a process that'd go down before we had the chance to get its app id (was retrieved post profiling).
* No chance for PID reuse collision - PID being reused by another app and we get the wrong data.

That's ^^ not true yet for pyperf, because pyperf still attaches this data post profiling. This will change in a PR after https://github.com/Granulate/gprofiler/pull/281 - where we will move the process selection to gprofiler from pyperf, and then gprofiler will take the metadata & id of apps before pyperf begins to profile them.

## How Has This Been Tested?
CI

### CODE REVIEW

**Review by adi-benz — COMMENTED** [review]
Looks great, and not such a big change LOC-wise

**Review by adi-benz — CHANGES_REQUESTED** [review]

File: gprofiler/gprofiler_types.py [hunk 1]
```diff
@@ -4,13 +4,32 @@
...
+class ProfileData:
+    """
+    Represents data collected by gProfiler about a process.
+    First and foremost are the stacks - the raw profiling data itself.
+    Then we have other "metadata"-ish fields like the appid and the app metadata, and more
+    will come.
+    """
+
+    stacks: StackToSampleCount
+    appid: Optional[str]
```
[the remark this answers is not in the document]
Comment (Jongy) [reply gprofiler/gprofiler_types.py:27]: Hmm also true, but let's not continue enlarging this PR, I'll do as a separate PR?

File: gprofiler/metadata/application_identifiers.py [hunk 2]
```diff
@@ -106,7 +106,7 @@ def gunicorn_to_application_name(self, wsgi_app_spec: str, process: Process) ->
 
 
 class _GunicornApplicationIdentifier(_GunicornApplicationIdentifierBase):
-    def get_application_name(self, process: Process) -> Optional[str]:
+    def get_app_id(self, process: Process) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (Jongy) [reply gprofiler/metadata/application_identifiers.py:109]: Yup

File: tests/conftest.py [hunk 3]
```diff
@@ -343,16 +344,17 @@ def assert_collapsed(runtime: str) -> AssertInCollapsed:
 
 @fixture
 def assert_application_name(application_pid: int, runtime: str, in_container: bool) -> Generator:
```
[the remark this answers is not in the document]
Comment (Jongy) [reply tests/conftest.py:346]: Fixed

File: tests/utils.py [hunk 4]
```diff
@@ -106,12 +106,20 @@ def assert_function_in_collapsed(function_name: str, collapsed: StackToSampleCou
...
     result = profiler.snapshot()
     assert len(result) == 1
     return next(iter(result.values()))
 
 
+def snapshot_one_collapsed(profiler: ProfilerInterface) -> StackToSampleCount:
+    return snapshot_one_profile(profiler).stacks
+
+
+def snapshot_pid_collaped(profiler: ProfilerInterface, pid: int) -> StackToSampleCount:
```
[the remark this answers is not in the document]
Comment (Jongy) [reply tests/utils.py:119]: Was fixed (refresh your view?)

[the remark this answers is not in the document]
Comment (Jongy) [reply tests/utils.py:119]: Or not. wtf

File: gprofiler/merge.py [hunk 5]
```diff
@@ -351,36 +350,36 @@ def concatenate_profiles(
...
             continue
 
-        process_perf = perf_pid_to_stacks_counter.get(pid)
+        process_perf = perf_pid_to_profiles.get(pid)
         if process_perf is None:
-            # no samples collected by perf for this process.
-            continue
+            # no samples collected by perf for this process, so those collected by the runtime profiler
+            # are dropped.
+            perf_samples_count = 0
```
[the remark this answers is not in the document]
Comment (Jongy) [reply gprofiler/merge.py:370]: `scale_sample_counts` won't write stacks that have 0 samples, so it actually ends up returning an empty `Counter`

**Review by adi-benz — DISMISSED** [review]

**Review by adi-benz — APPROVED** [review]

**Pull request merged**